ls1intum/Artemis
# `General`: Improve the appearance of system notifications

### Checklist
#### General


#### Server
#### Client

### Motivation and Context



System notifications are shown on top of the page on all clients, e.g. to inform about an upcoming maintenance period.
Currently, they look like our alerts looked previously, and they are not working properly at the moment:
- Only one notification is supported at the same time
- Updating the shown notifications via websocket was not implemented properly, and often a reload of the app was required to make it show up.
- The management interface nor the server side endpoints did do any proper user input validation

![grafik](https://user-images.githubusercontent.com/23171488/180484643-82eca760-f0ce-402e-bf57-cf9b82976b3d.png)

### Description


I've decided to give it a full overhaul.
The following changes have been added:

- Changed appearance of the notifications: They are now glued to the bottom of the navbar and look more modern; they also got a intro animation as moving things usually raise a users attention more effectively that a colorful, but static item
- Notifications are now temporarily dismissable, but will show up again on a reload or system notification update
- Notifications with a date in the future will now show up exactly in time without a reload
- The management view for system notifications has been updated:
  * Added three notification states: Scheduled, active, and expired (previously: Active, Not Active)
  * The list of notifications is now sorted by notification date by default
  * The update component now validates that the type is not empty, and that the expiration date is after the notification date

### Steps for Testing

Prerequisites:
- 1 Admin

1. Log in to Artemis
2. Navigate to Server Administration > System notification
3. Create multiple new system notifications. Play around with the types, the title, the text (which is optiona), and the dates: Make them active right after creating them; make one active a minute in the future for example and check that it shows up in time; make one that expires in a minute and check that it actually disappears in time ....... etc)
4. Make sure that you can not set an expiration date that is before the notification date. You can even try to modify the request, the server should check it as well
5. Verify that the "Active", "Scheduled" and "Expirted" batches are shown correctly and update live as well
6. Verify that you can temporatily dismiss a notification and it shows up again after a reload
7. Verify that all dismissed notifications show up again when you update a notification
8. Log out, check that the notifications are visible if you're not logged in as well (after a fresh reload)
9. Check that the notifications wrap correctly on mobile phones: The title and the icon should move into the first line, the text below it

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| system-notification.component.ts | 70.83 | 94.82 |
| system-notification-management-update.component.ts | 86.25% | 88.88% |
| system-notification-management.component.ts | 70.37% | 84.81% |

### Screenshots



![Screenshot 2022-07-22 at 18 48 01](https://user-images.githubusercontent.com/23171488/180486565-c25e4023-036a-4c61-844b-4b3f53ab0373.png)
![Screenshot 2022-07-22 at 17 27 05](https://user-images.githubusercontent.com/23171488/180486568-541ecdf5-9dac-4008-ba9e-319fe7b645e8.png)
